### PR TITLE
:seedling: Add OKTA_MFA_CHOICE in config.properties

### LINKED ISSUE #258: Add OKTA_MFA_CHOICE in config.properties
See #128 for additional context.

The last thing that requires user-input during a login is choosing the MFA option. I would like to be able to specify it in `config.properties` instead, so the login can be fully non-interactive.

The value could be [PROVIDER_TYPE][].[FACTOR_TYPE][]. For example:

```
OKTA_MFA_CHOICE=OKTA.push
```

[provider_type]: https://developer.okta.com/docs/api/resources/factors#provider-type
[factor_type]: https://developer.okta.com/docs/api/resources/factors#factor-type

### PR DESCRIPTION
**Pull request by AlainODea**
Problem Statement
-----------------
Issue #258 states:
> See #128 for additional context.
> 
> The last thing that requires user-input during a login is choosing the MFA option. I would like to be able to specify it in `config.properties` instead, so the login can be fully non-interactive.
> 
> The value could be [PROVIDER_TYPE](https://developer.okta.com/docs/api/resources/factors#provider-type).[FACTOR_TYPE](https://developer.okta.com/docs/api/resources/factors#factor-type). For example:
> 
> ```
> OKTA_MFA_CHOICE=OKTA.push
> ```

Solution
--------
 - Allow MFA choice to be specified as provider and factor type to use

 - Exs: OKTA.token:software:totp, GOOGLE.token:software:totp, OTKA.push

Resolves #258

### CODE REVIEW

**Comment by AlainODea** [comment]
@pbrisbin please review.

Most of the code damage here is dependency inversion stuff.

There is a debugging fix (Scanner.next() should be Scanner.nextLine()) weaved in there as well to make it possible to use passwords with spaces when debugging in the IDE.

File: src/main/java/com/okta/tools/authentication/OktaMFA.java [hunk 1]
```diff
@@ -120,13 +126,22 @@ private static JSONObject selectFactor(JSONObject primaryAuthResponse) throws JS
...
         if (supportedFactors.size() > 1) {
+            if (environment.oktaMfaChoice != null) {
+                for (JSONObject factor : supportedFactors) {
+                    String provider = factor.getString("provider");
+                    String factorType = factor.getString("factorType");
+                    if ((provider + "." + factorType).equals(environment.oktaMfaChoice)) {
+                        return factor;
+                    }
+                }
+            }
```
Comment (AlainODea) [inline src/main/java/com/okta/tools/authentication/OktaMFA.java]: :seedling: core of the change. Halt MFA menu presentation and take the option based on config.properties/envvars if matches.

File: src/main/java/com/okta/tools/authentication/OktaMFA.java [hunk 2]
```diff
@@ -120,13 +126,22 @@ private static JSONObject selectFactor(JSONObject primaryAuthResponse) throws JS
...
             System.err.println("\nMulti-Factor authentication is required. Please select a factor to use.");
             System.err.println("Factors:");
             for (int i = 0; i < supportedFactors.size(); i++) {
                 JSONObject factor = supportedFactors.get(i);
-                factorType = factor.getString("factorType");
-                factorType = getFactorDescription(factorType, factor);
-                System.err.println("[ " + (i + 1) + " ] : " + factorType);
+                String factorType = factor.getString("factorType");
+                String factorDescription = getFactorDescription(factorType, factor);
+                System.err.println("[ " + (i + 1) + " ] : " + factorDescription);
```
Comment (AlainODea) [inline src/main/java/com/okta/tools/authentication/OktaMFA.java]: :art: moved scope of variables inward and gave distinct states different variable names.

File: Readme.MD [hunk 3]
```diff
@@ -159,6 +159,7 @@ Here is the list of parameters that can be environment variables or settings in
   - ```OKTA_AWS_REGION``` is the default AWS region to store with the created profile.
   - ```OKTA_AWS_ROLE_TO_ASSUME``` is the IAM Role ARN to use. If present will try to match okta account's retrieved role list and use it. Will still prompt if no match found. (ex: **arn:aws:iam::123456789012:role/EC2-Admins**)
   - ```OKTA_STS_DURATION``` is the duration the role will be assumed, in seconds. The maximum session duration allowed by AWS is 12 hours and this needs to be set on the role as well.  Defaults to 1hr.
+  - ```OKTA_MFA_CHOICE``` is the provider and factor type to use if prompted for MFA.  Example: ```OKTA.push```. See [Factors documentation](https://developer.okta.com/docs/api/resources/factors#factor-type) for values. (default: use single factor or prompt user to select from usable factors).
```
Comment (AlainODea) [inline Readme.MD]: I'm open to rewording this. It is a first pass.

File: src/main/java/com/okta/tools/ListRoles.java [hunk 4]
```diff
@@ -16,21 +16,22 @@
...
-import java.util.Map;
 
 public class ListRoles {
     public static void main(String[] args) throws Exception {
         OktaAwsCliEnvironment environment = OktaAwsConfig.loadEnvironment();
         CookieHelper cookieHelper = new CookieHelper(environment);
-        OktaSaml oktaSaml = new OktaSaml(environment, cookieHelper);
+        OktaMFA oktaMFA = new OktaMFA(environment);
+        OktaAuthentication oktaAuthentication = new OktaAuthentication(environment, oktaMFA);
+        OktaSaml oktaSaml = new OktaSaml(environment, cookieHelper, oktaAuthentication);
```
Comment (AlainODea) [inline src/main/java/com/okta/tools/ListRoles.java]: :art: dependency inversion refactor junk.

File: src/main/java/com/okta/tools/OktaAwsCliAssumeRole.java [hunk 5]
```diff
@@ -62,8 +64,10 @@ private void init() throws Exception {
         roleHelper = new RoleHelper(environment);
         credentialsHelper  = new CredentialsHelper(environment);
         profileHelper = new ProfileHelper(credentialsHelper, environment);
+        OktaMFA oktaMFA = new OktaMFA(environment);
+        OktaAuthentication oktaAuthentication = new OktaAuthentication(environment, oktaMFA);
 
-        oktaSaml = new OktaSaml(environment, cookieHelper);
+        oktaSaml = new OktaSaml(environment, cookieHelper, oktaAuthentication);
```
Comment (AlainODea) [inline src/main/java/com/okta/tools/OktaAwsCliAssumeRole.java]: :art: dependency inversion refactor junk.

File: src/main/java/com/okta/tools/OktaAwsCliEnvironment.java [hunk 6]
```diff
@@ -38,6 +39,7 @@ public OktaAwsCliEnvironment(boolean browserAuth, String oktaOrg,
         this.awsRoleToAssume = awsRoleToAssume;
         this.stsDuration = stsDuration;
         this.awsRegion = awsRegion;
+        this.oktaMfaChoice = oktaMfaChoice;
```
Comment (AlainODea) [inline src/main/java/com/okta/tools/OktaAwsCliEnvironment.java]: :art: dependency inversion refactor junk.

File: src/main/java/com/okta/tools/OktaAwsConfig.java [hunk 7]
```diff
@@ -57,6 +57,7 @@ static OktaAwsCliEnvironment loadEnvironment(String profile) {
                 getEnvOrConfig(properties, "OKTA_AWS_ROLE_TO_ASSUME"),
                 getStsDurationOrDefault(getEnvOrConfig(properties, "OKTA_STS_DURATION")),
                 getAwsRegionOrDefault(getEnvOrConfig(properties, "OKTA_AWS_REGION")),
+                getEnvOrConfig(properties, "OKTA_MFA_CHOICE"),
```
Comment (AlainODea) [inline src/main/java/com/okta/tools/OktaAwsConfig.java]: :seedling: config.properties and env var support for OKTA_MFA_CHOICE.

File: src/main/java/com/okta/tools/saml/OktaSaml.java [hunk 8]
```diff
@@ -36,15 +36,15 @@
 
     private final OktaAwsCliEnvironment environment;
     private final CookieHelper cookieHelper;
+    private final OktaAuthentication authentication;
 
-    public OktaSaml(OktaAwsCliEnvironment environment, CookieHelper cookieHelper) {
+    public OktaSaml(OktaAwsCliEnvironment environment, CookieHelper cookieHelper, OktaAuthentication oktaAuthentication) {
         this.environment = environment;
         this.cookieHelper = cookieHelper;
+        this.authentication = oktaAuthentication;
```
Comment (AlainODea) [inline src/main/java/com/okta/tools/saml/OktaSaml.java]: :art: dependency inversion refactor junk.

**Comment by AlainODea** [comment]
@pbrisbin do you have anyone to recommend? It's difficult to find code reviewers for this.

**Comment by AlainODea** [comment]
@eborden sure thing.

**Comment by AlainODea** [comment]
This codebase is very not test-friendly 😢  This is going to be rough.

**Comment by eborden** [comment]
I wouldn't struggle too long with it. Adding DI moves it in the direction of testable, so it is at least an incremental improvement.

**Comment by AlainODea** [comment]
@eborden I think the added tests in 8685099bd58df9328e2a8848beb4b80498f6f40f should address the test coverage for this feature and improve overall coverage marginally.

**Comment by AlainODea** [comment]
My pleasure, thank you for the code review ans keeping me honest 😅

**Pull request merged**